### PR TITLE
Cross-Chain Queries (ICS 031)

### PR DESCRIPTION
**Pull request by SangyunOck**
Closes: #2707 

## Description

* Extra pr for maintenance of ics-031 implementation (https://github.com/informalsystems/hermes/pull/2709).
* Based on hermes test-framework(https://github.com/informalsystems/hermes/tree/master/tools/test-framework). 
* Compiled chain binary included(GNU/Linux, arm64) which is automatically chosen on runtime.

The only difference between last pr is addition of "tools/integration-test/src/tests/test_ics_031_crosschain_query_implementation_2709.rs"

Thanks!

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Comment by sampocs** [comment]
Excellent work here! We've been using ICQ queries in production (with the go relayer), so I thought I would share my thoughts!

The two things that jumped out to me were: (and it’s totally possible I might have just misinterpreted the implementation)
1. I didn’t see proofs attached to the queries - did I just miss it or is it being verified in some other way? Need to prevent the case where a relayer submits a malicious response
2. It looked like the query type is defined using the REST API endpoint - does that mean you have to specify a specific node for each query? Also, is there a way to do it with the RPC endpoint instead? Otherwise, there’d now a dependency on 4 endpoints for hermes

**Comment by sampocs** [comment]
All good points! I agree the abci_query is quite limiting! That came to a surprise for us - it would really be great to be able to query any endpoint that's available with REST (e.g. querying for data stored in a list). However, the proof does feel mandatory. If only there was a way to get both!

File: tools/test-framework/src/relayer/chain.rs [hunk 1]
```diff
@@ -1,24 +1,24 @@
 /*!
-   Definition for a proxy [`ChainHandle`] implementation for tagged
```
Comment (romac) [inline tools/test-framework/src/relayer/chain.rs:2]: Can we undo the whitespace changes here please?

File: crates/relayer/src/worker/map.rs [hunk 2]
```diff
@@ -237,5 +237,6 @@ fn metric_type(o: &Object) -> ibc_telemetry::state::WorkerType {
         Object::Channel(_) => WorkerType::Channel,
         Object::Packet(_) => WorkerType::Packet,
         Object::Wallet(_) => WorkerType::Wallet,
+        Object::CrossChainQueryPacket(_) => WorkerType::CrossChainQueryPacket,
```
Comment (romac) [inline crates/relayer/src/worker/map.rs:240]: Let's rename `Object::CrossChainQueryPacket` to `Object::CrossChainQuery` and `WorkerType::CrossChainQueryPacket` to `WorkerType::CrossChainQuery`

File: crates/telemetry/src/state.rs [hunk 3]
```diff
@@ -71,6 +72,7 @@ impl Display for WorkerType {
             Self::Channel => write!(f, "channel"),
             Self::Packet => write!(f, "packet"),
             Self::Wallet => write!(f, "wallet"),
+            Self::CrossChainQueryPacket => write!(f, "cross_chain_query_packet"),
```
Comment (romac) [inline crates/telemetry/src/state.rs:75]: Here as well:

```suggestion
            Self::CrossChainQuery => write!(f, "cross_chain_query"),
```

File: tools/integration-test/src/tests/mod.rs [hunk 4]
```diff
@@ -21,6 +21,8 @@ pub mod tendermint;
 pub mod ternary_transfer;
 pub mod transfer;
 
+pub mod test_ics_031_crosschain_query_implementation_2709;
```
Comment (romac) [inline tools/integration-test/src/tests/mod.rs:24]:
```suggestion
pub mod crosschain_query;
```

File: crates/relayer/Cargo.toml [hunk 5]
```diff
@@ -61,6 +61,7 @@ humantime = "2.1.0"
 regex = "1.5.5"
 moka = "0.9.4"
 uuid = { version = "1.1.2", features = ["v4"] }
+reqwest = "0.11.11"
```
Comment (romac) [inline crates/relayer/Cargo.toml:64]: Why is this needed?

File: crates/relayer-cli/src/cli_utils.rs [hunk 6]
```diff
@@ -71,7 +71,8 @@ pub fn spawn_chain_runtime_generic<Handle: ChainHandle>(
     chain_id: &ChainId,
 ) -> Result<Handle, Error> {
     let rt = Arc::new(TokioRuntime::new().unwrap());
-    spawn::spawn_chain_runtime(config, chain_id, rt).map_err(Error::spawn)
+    let query_rt = Arc::new(TokioRuntime::new().unwrap());
```
Comment (romac) [inline crates/relayer-cli/src/cli_utils.rs:74]: Why do we need another runtime here?

File: crates/modules/src/applications/ics31_cross_chain_query/events.rs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+/// Types for the IBC events emitted from Tendermint Websocket by the cross-chain-query module.
+
+pub const ATTRIBUTE_CHAIN_ID_KEY: &str = "chain_id";
+pub const ATTRIBUTE_QUERY_HEIGHT_KEY: &str = "query_height";
+pub const ATTRIBUTE_QUERY_ID_KEY: &str = "query_id";
+pub const ATTRIBUTE_QUERY_PATH_KEY: &str = "query_path";
+
+#[derive(Serialize, Clone, Debug, PartialEq, Eq)]
+pub struct SendPacket {
```
Comment (romac) [inline crates/modules/src/applications/ics31_cross_chain_query/events.rs:19]: Why is this needed, rather than just use `CrossChainQueryPacket` directly?

File: .changelog/unreleased/features/ibc-relayer/2707-ics-031-implementation.md [hunk 8]
```diff
@@ -0,0 +1 @@
+- ics-031-implementation for ibc-rs ([#2707](https://github.com/informalsystems/ibc-rs/issues/2707))
```
Comment (romac) [inline .changelog/unreleased/features/ibc-relayer/2707-ics-031-implementation.md:1]:
```suggestion
- Add support for Cross-Chain Queries to the relayer (ICS 031)
  ([#2707](https://github.com/informalsystems/ibc-rs/issues/2707))
```

File: crates/relayer/src/event/monitor.rs [hunk 9]
```diff
@@ -134,15 +138,19 @@ pub mod queries {
...
 }
 
 impl EventMonitor {
     /// Create an event monitor, and connect to a node
     #[instrument(
-        name = "event_monitor.create",
-        level = "error",
-        skip_all,
-        fields(chain = %chain_id, addr = %node_addr)
+    name = "event_monitor.create",
```
Comment (romac) [inline crates/relayer/src/event/monitor.rs:150]: Can we undo those whitespace changes please?

File: crates/relayer/src/event/monitor.rs [hunk 10]
```diff
@@ -186,7 +194,7 @@ impl EventMonitor {
     }
 
     /// Clear the current subscriptions, and subscribe again to all queries.
-    #[instrument(name = "event_monitor.subscribe", skip_all, fields(chain = %self.chain_id))]
+    #[instrument(name = "event_monitor.subscribe", skip_all, fields(chain = % self.chain_id))]
```
Comment (romac) [inline crates/relayer/src/event/monitor.rs:197]: Same here

File: crates/relayer/src/chain/cosmos/query/custom_query.rs [hunk 11]
```diff
@@ -0,0 +1,52 @@
+use crate::chain::requests::CrossChainQueryRequest;
+use crate::chain::responses::CrossChainQueryResponse;
+use reqwest::{Client, Error};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct MsgTransfer {
```
Comment (romac) [inline crates/relayer/src/chain/cosmos/query/custom_query.rs:7]: What's the purpose of this definition?

File: crates/relayer/src/chain/cosmos/query/custom_query.rs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    pub amount: String,
+    pub denom: String,
+    pub receiver: String,
+    pub sender: String,
+}
+
+pub async fn rest_query(
```
Comment (romac) [inline crates/relayer/src/chain/cosmos/query/custom_query.rs:14]: Can we use gRPC instead?

File: crates/relayer/src/link/error.rs [hunk 13]
```diff
@@ -32,6 +32,10 @@ define_error! {
                 format!("failed to construct packet proofs for chain {0}", e.chain_id)
             },
 
+        LocalBound
+            [ Error ]
+            |_| {" local bound error "},
```
Comment (romac) [inline crates/relayer/src/link/error.rs:37]:
```suggestion
            |_| { "local bound error "},
```

File: crates/relayer/src/link/relay_path.rs [hunk 14]
```diff
@@ -1430,6 +1434,32 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
         Ok(unprocessed)
     }
 
+    // TODO: handle cross-chain query
```
Comment (romac) [inline crates/relayer/src/link/relay_path.rs:1437]: There is a TODO left here

File: crates/relayer/src/object.rs [hunk 15]
```diff
@@ -282,6 +308,8 @@ impl Object {
             Self::Channel(ref channel) => &channel.dst_chain_id,
             Self::Packet(ref path) => &path.dst_chain_id,
             Self::Wallet(ref wallet) => &wallet.chain_id,
+            // TODO: set to dst chain id when needed
```
Comment (romac) [inline crates/relayer/src/object.rs:311]: There's a TODO left here

File: crates/relayer/src/supervisor.rs [hunk 16]
```diff
@@ -55,11 +59,11 @@ type ArcBatch = Arc<monitor::Result<EventBatch>>;
 type Subscription = Receiver<ArcBatch>;
 
 /**
-    A wrapper around the SupervisorCmd sender so that we can
-    send stop signal to the supervisor before stopping the
-    chain drivers to prevent the supervisor from raising
-    errors caused by closed connections.
-*/
+A wrapper around the SupervisorCmd sender so that we can
```
Comment (romac) [inline crates/relayer/src/supervisor.rs:62]: Can we undo these whitespace changes please?

File: crates/relayer/src/chain/cosmos.rs [hunk 17]
```diff
@@ -125,9 +128,11 @@ pub struct CosmosSdkChain {
     config: ChainConfig,
     tx_config: TxConfig,
     rpc_client: HttpClient,
+    rest_client: RestClient,
     grpc_addr: Uri,
     light_client: TmLightClient,
     rt: Arc<TokioRuntime>,
+    query_rt: Arc<TokioRuntime>,
```
Comment (romac) [inline crates/relayer/src/chain/cosmos.rs:135]: Do we really need another Tokio runtime?

File: crates/relayer/src/chain/cosmos.rs [hunk 18]
```diff
@@ -125,9 +128,11 @@ pub struct CosmosSdkChain {
     config: ChainConfig,
     tx_config: TxConfig,
     rpc_client: HttpClient,
+    rest_client: RestClient,
```
Comment (romac) [inline crates/relayer/src/chain/cosmos.rs:131]: Can we use the gRPC client instead?

File: crates/relayer/src/chain/cosmos.rs [hunk 19]
```diff
@@ -618,54 +631,12 @@ impl ChainEndpoint for CosmosSdkChain {
...
-    /// TODO - more work is required here for a smarter split maybe iteratively accumulating/ evaluating
-    /// msgs in a Tx until any of the max size, max num msgs, max fee are exceeded.
-    fn send_messages_and_wait_commit(
-        &mut self,
-        tracked_msgs: TrackedMsgs,
-    ) -> Result<Vec<IbcEventWithHeight>, Error> {
-        let runtime = self.rt.clone();
-
-        runtime.block_on(self.do_send_messages_and_wait_commit(tracked_msgs))
+    fn keybase(&self) -> &KeyRing {
```
Comment (romac) [inline crates/relayer/src/chain/cosmos.rs:634]: Can we undo the re-ordering of these methods, please?

**Comment by SangyunOck** [comment]
Hi @sampocs, 
I included ProofOps in this pr: https://github.com/informalsystems/hermes/pull/2915
Would you please give me an e-mail? I have some question about Stride :)
